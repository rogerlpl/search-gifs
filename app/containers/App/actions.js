/*
 * App Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
  LOAD_GIFS,
  LOAD_GIFS_SUCCESS,
  LOAD_GIFS_ERROR,
} from './constants';

/**
 * Load the repositories, this action starts the request saga
 *
 * @return {object} An action object with a type of LOAD_GIFS
 */
export function loadGifs() {
  return {
    type: LOAD_GIFS,
  };
}

/**
re loaded by the request saga * Dispatched when the gifs a
 *
 * @param  {array} gifs The gifs data
 * @param  {string} keyword The current keyword
 *
 * @return {object}      An action object with a type of LOAD_GIFS_SUCCESS passing the gifs
 */
export function gifsLoaded(gifs, keyword) {
  return {
    type: LOAD_GIFS_SUCCESS,
    gifs,
    keyword,
  };
}

/**
 * Dispatched when loading the gifs fails
 *
 * @param  {object} error The error
 *
 * @return {object}       An action object with a type of LOAD_GIFS_ERROR passing the error
 */
export function gifLoadingError(error) {
  return {
    type: LOAD_GIFS_ERROR,
    error,
  };
}
