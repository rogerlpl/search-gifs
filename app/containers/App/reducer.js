/*
 * AppReducer
 *
 * The reducer takes care of our data. Using actions, we can change our
 * application state.
 * To add a new action, add it to the switch statement in the reducer function
 *
 * Example:
 * case YOUR_ACTION_CONSTANT:
 *   return state.set('yourStateVariable', true);
 */

import { fromJS } from 'immutable';

import {
  LOAD_GIFS_SUCCESS,
  LOAD_GIFS,
  LOAD_GIFS_ERROR,
} from './constants';

// The initial state of the App
const initialState = fromJS({
  loading: false,
  error: false,
  currentKeyword: false,
  gifData: {
    gifs: false,
  },
});

function appReducer(state = initialState, action) {
  switch (action.type) {
    case LOAD_GIFS:
      return state
        .set('loading', true)
        .set('error', false)
        .setIn(['gifData', 'gifs'], false);
    case LOAD_GIFS_SUCCESS:
      return state
        .setIn(['gifData', 'gifs'], action.gifs)
        .set('loading', false)
        .set('currentKeyword', action.keyword);
    case LOAD_GIFS_ERROR:
      return state
        .set('error', action.error)
        .set('loading', false);
    default:
      return state;
  }
}

export default appReducer;
