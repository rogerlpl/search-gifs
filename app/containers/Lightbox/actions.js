/*
 * Lightbox Actions
 *
 * Actions change things in your application
 * Since this boilerplate uses a uni-directional data flow, specifically redux,
 * we have these actions which are the only way your application interacts with
 * your application state. This guarantees that your state is up to date and nobody
 * messes it up weirdly somewhere.
 *
 * To add a new Action:
 * 1) Import your constant
 * 2) Add a function like this:
 *    export function yourAction(var) {
 *        return { type: YOUR_ACTION_CONSTANT, var: var }
 *    }
 */

import {
    OPEN_LIGHTBOX,
    CLOSE_LIGHTBOX,
    SET_SLIDE 
} from './constants';

/**
 * Set the current slide
 *
 * @param  {number} slide The number of the new current slide
 *
 * @return {object}    An action object with a type of CHANGE_KEYWORD
 */
export function setCurrentSlide(slide) {
  return {
    type: SET_SLIDE,
    slide
  };
}
/**
 * Open the Lightbox Modal
 *
 * @return {object}    An action object with a type of CHANGE_KEYWORD
 */
export function openLightbox() {
  return {
    type: OPEN_LIGHTBOX,
  };
}
/**
 * Close the Lightbox Modal
 *
 * @return {object}    An action object with a type of CHANGE_KEYWORD
 */
export function closeLightbox() {
  return {
    type: CLOSE_LIGHTBOX,
  };
}
