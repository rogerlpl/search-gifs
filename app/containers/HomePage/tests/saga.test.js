/**
 * Tests for HomePage sagas
 */

import { put, takeLatest } from 'redux-saga/effects';

import { LOAD_GIFS } from 'containers/App/constants';
import { gifsLoaded, gifLoadingError } from 'containers/App/actions';

import giphyData, { getGifs } from '../saga';

const keyword = 'cats';

/* eslint-disable redux-saga/yield-effects */
describe('getGifs Saga', () => {
  let getGifsGenerator;

  // We have to test twice, once for a successful load and once for an unsuccessful one
  // so we do all the stuff that happens beforehand automatically in the beforeEach
  beforeEach(() => {
    getGifsGenerator = getGifs();

    const selectDescriptor = getGifsGenerator.next().value;
    expect(selectDescriptor).toMatchSnapshot();

    const callDescriptor = getGifsGenerator.next(keyword).value;
    expect(callDescriptor).toMatchSnapshot();
  });

  it('should dispatch the gifsLoaded action if it requests the data successfully', () => {
    const response = [{
      name: 'First gif',
    }, {
      name: 'Second gif',
    }];
    const putDescriptor = getGifsGenerator.next(response).value;
    expect(putDescriptor).toEqual(put(gifsLoaded(response, keyword)));
  });

  it('should call the gifLoadingError action if the response errors', () => {
    const response = new Error('Some error');
    const putDescriptor = getGifsGenerator.throw(response).value;
    expect(putDescriptor).toEqual(put(gifLoadingError(response)));
  });
});

describe('giphyDataSaga Saga', () => {
  const giphyDataSaga = giphyData();

  it('should start task to watch for LOAD_GIFS action', () => {
    const takeLatestDescriptor = giphyDataSaga.next().value;
    expect(takeLatestDescriptor).toEqual(takeLatest(LOAD_GIFS, getGifs));
  });
});
