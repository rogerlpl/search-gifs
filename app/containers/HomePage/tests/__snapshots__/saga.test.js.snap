// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getGifs Saga should call the gifLoadingError action if the response errors 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getGifs Saga should call the gifLoadingError action if the response errors 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://api.giphy.com/v1/gifs/search?api_key=Z5XiFHOYMn8y8muarg2Bg8PCmpxQZyWz&q=cats",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;

exports[`getGifs Saga should dispatch the gifsLoaded action if it requests the data successfully 1`] = `
Object {
  "@@redux-saga/IO": true,
  "SELECT": Object {
    "args": Array [],
    "selector": [Function],
  },
}
`;

exports[`getGifs Saga should dispatch the gifsLoaded action if it requests the data successfully 2`] = `
Object {
  "@@redux-saga/IO": true,
  "CALL": Object {
    "args": Array [
      "https://api.giphy.com/v1/gifs/search?api_key=Z5XiFHOYMn8y8muarg2Bg8PCmpxQZyWz&q=cats",
    ],
    "context": null,
    "fn": [Function],
  },
}
`;
